src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -163,6 +168,76 @@ protected ExecutionStrategy(DataFetcherExceptionHandler dataFetcherExceptionHand
+        ExecutionStrategyInstrumentationContext executionStrategyCtx = ExecutionStrategyInstrumentationContext.nonNullCtx(instrumentation.beginExecutionStrategy(instrumentationParameters, executionContext.getInstrumentationState()));
+
+        MergedSelectionSet fields = parameters.getFields();
+        List<String> fieldNames = fields.getKeys();
+        Async.CombinedBuilder<FieldValueInfo> futures = resolveFieldsWithInfo(executionContext, parameters, fields, fieldNames);
+
+        CompletableFuture<Object> overallResult = new CompletableFuture<>();
+        CompletableFuture<ExecutionResult> instrumentationFuture = overallResult.thenApply(val -> ExecutionResult.newExecutionResult().data(val).build());
+        executionStrategyCtx.onDispatched(instrumentationFuture);
+
+        futures.await().whenComplete((completeValueInfos, throwable) -> {
+            BiConsumer<List<Object>, Throwable> handleResultsConsumer = handleFieldResults(executionContext, fieldNames, overallResult);
+            if (throwable != null) {
+                handleResultsConsumer.accept(null, throwable.getCause());
+                return;
+            }
+            Async.CombinedBuilder<Object> resultFutures = Async.ofExpectedSize(completeValueInfos.size());
+            for (FieldValueInfo completeValueInfo : completeValueInfos) {
+                resultFutures.add(completeValueInfo.getValue());
+            }
+            executionStrategyCtx.onFieldValuesInfo(completeValueInfos);
+            resultFutures.await().whenComplete(handleResultsConsumer);
+        }).exceptionally((ex) -> {
+            // if there are any issues with combining/handling the field results,
+            // complete the future at all costs and bubble up any thrown exception so
+            // the execution does not hang.
+            executionStrategyCtx.onFieldValuesException();
+            overallResult.completeExceptionally(ex);
+            return null;
+        });
+
+        instrumentationFuture.whenComplete(executionStrategyCtx::onCompleted);
+        return overallResult;
+    }
+
+    @NotNull
+    protected Async.CombinedBuilder<FieldValueInfo> resolveFieldsWithInfo(ExecutionContext executionContext, ExecutionStrategyParameters parameters, MergedSelectionSet fields, List<String> fieldNames) {
+        Async.CombinedBuilder<FieldValueInfo> futures = Async.ofExpectedSize(fields.size());
+        for (String fieldName : fieldNames) {
+            MergedField currentField = fields.getSubField(fieldName);
+
+            ResultPath fieldPath = parameters.getPath().segment(mkNameForPath(currentField));
+            ExecutionStrategyParameters newParameters = parameters
+                    .transform(builder -> builder.field(currentField).path(fieldPath).parent(parameters));
+
+            CompletableFuture<FieldValueInfo> future = resolveFieldWithInfo(executionContext, newParameters);
+            futures.add(future);
+        }
+        return futures;
+    }
```

Comment:
This is a common method extractred since `execute` and `executeObject` use this pattern